assets/src/components/routeLadder.tsx

```diff
@@ -157,6 +159,79 @@ export const groupByPosition = (
       incoming: [],
     } as ByPosition
   )
+
+  const incomingGhosts: Ghost[] = [
+    ...generateVirtualGhostsForLateIncomingVehicles(realVehicles.incoming),
+    ...generateVirtualGhostsForLateLayingOverVehicles([
+      ...realVehicles.layingOverTop,
+      ...realVehicles.layingOverBottom,
+    ]),
+  ]
+
+  return {
+    ...realVehicles,
+    onRoute: [...realVehicles.onRoute, ...incomingGhosts],
+  }
 }
 
+const generateVirtualGhostsForLateIncomingVehicles = (
+  incomingVehiclesOrGhosts: VehicleOrGhost[]
+): Ghost[] =>
+  lateStartingVehicles(incomingVehiclesOrGhosts).map(vehicle =>
+    ghostFromVehicleScheduledLocation(vehicle)
+  )
+
+const generateVirtualGhostsForLateLayingOverVehicles = (
+  incomingVehiclesOrGhosts: VehicleOrGhost[]
+): Ghost[] =>
+  lateStartingLayingOverVehicles(incomingVehiclesOrGhosts).map(vehicle =>
+    ghostFromVehicleScheduledLocation(vehicle)
+  )
+
+const lateStartingVehicles = (
+  incomingVehiclesOrGhosts: VehicleOrGhost[]
+): Vehicle[] =>
+  incomingVehiclesOrGhosts.filter(vehicleOrGhost =>
+    isAVehicleThatIsLateStartingScheduledTrip(vehicleOrGhost)
+  ) as Vehicle[]
+
+const lateStartingLayingOverVehicles = (
+  layingOverVehiclesOrGhosts: VehicleOrGhost[]
+): Vehicle[] =>
+  layingOverVehiclesOrGhosts.filter(
+    vehicleOrGhost =>
+      isAVehicleThatIsLateStartingScheduledTrip(vehicleOrGhost) &&
+      isScheduledForCurrentDirection(vehicleOrGhost as Vehicle)
+  ) as Vehicle[]
+
+const isAVehicleThatIsLateStartingScheduledTrip = (
+  vehicleOrGhost: VehicleOrGhost
+): boolean =>
+  isVehicle(vehicleOrGhost) &&
+  hasAScheduleLocation(vehicleOrGhost) &&
+  isLateStartingScheduledTrip(vehicleOrGhost)
+
+const isScheduledForCurrentDirection = (vehicle: Vehicle): boolean =>
+  vehicle.directionId === vehicle.scheduledLocation!.directionId
+
+const hasAScheduleLocation = (vehicle: Vehicle): boolean =>
```

Comment:
I hear what you are saying about the size of the file, but I disagree about the value of these functions. Personally I find it much easier to follow logic when I'm able to read words rather than decipher code. Maybe is there a way you can imagine refactoring this file to break it apart, or another improvement along those lines?